### ISSUE TITLE
rootio: add support for TChain

### ISSUE THREAD

**Issue by sbinet** [go-hep/hep#44]
it should be possible to combine multiple `rootio.Tree`s (with the same layout) from multiple ROOT files into one `rootio.Tree`.

right, now, to get at a `rootio.Tree`, one does:

```go
func main() {
	f, err := rootio.Open("testdata/chain.1.root")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	obj, err := f.Get("tree")
	if err != nil {
		log.Fatal(err)
	}
	tree := obj.(rootio.Tree)
	fmt.Printf("entries= %v\n", tree.Entries())
}
```

the `rootio.Tree` interface is:
```go
// Tree is a collection of branches of data.
type Tree interface {
	Named
	Entries() int64
	TotBytes() int64
	ZipBytes() int64
	Branch(name string) Branch
	Branches() []Branch
	Leaves() []Leaf

	getFile() *File
	loadEntry(i int64) error
}
```

it should be possible to have a type that implements the `rootio.Tree` interface but logically concatenates multiple `rootio.Tree`s together.
Something like the `io.MultiReader` function: https://godoc.org/io#MultiReader

Initially, one could just require the user to pass already created/retrieved `rootio.Tree` to a, _e.g._, `Chain` function:

```go
// Chain returns a new Tree that is the logical concatenation of all the input Trees.
func Chain(trees ...Tree) Tree { ... }
```

one should think about resource ownership:
- who should close the ROOT file containing a Tree when that Tree has been completely read ?
- how to integrate the `chain` type with the `Scanner`s ?
- is it better to have a `Chain` function that takes a tree name and a list of file names ? (so resource ownership is clearer). it's less composable...
- perhaps introduce a `ChainScanner` that opens and closes files as needed when each `Tree` has been consumed ?

**Comment by sbinet** [comment]
@maloft, all yours :)

**Comment by sbinet** [comment]
f799e0b added 2 test files:
- `testdata/chain.1.root`
- `testdata/chain.2.root`

```
$> root-dump ./testdata/chain*
>>> file[./testdata/chain.1.root]
key[000]: tree;1 "my tree title" (TTree)
[000][evt]: {beg-000 0 [0 0 0 0 0 0 0 0 0 0] 0 [] std-000 [] [] end-000}
[001][evt]: {beg-001 1 [1 1 1 1 1 1 1 1 1 1] 1 [1] std-001 [1] [vec-001] end-001}
[002][evt]: {beg-002 2 [2 2 2 2 2 2 2 2 2 2] 2 [2 2] std-002 [2 2] [vec-002 vec-002] end-002}
[003][evt]: {beg-003 3 [3 3 3 3 3 3 3 3 3 3] 3 [3 3 3] std-003 [3 3 3] [vec-003 vec-003 vec-003] end-003}
[004][evt]: {beg-004 4 [4 4 4 4 4 4 4 4 4 4] 4 [4 4 4 4] std-004 [4 4 4 4] [vec-004 vec-004 vec-004 vec-004] end-004}
[005][evt]: {beg-005 5 [5 5 5 5 5 5 5 5 5 5] 5 [5 5 5 5 5] std-005 [5 5 5 5 5] [vec-005 vec-005 vec-005 vec-005 vec-005] end-005}
[006][evt]: {beg-006 6 [6 6 6 6 6 6 6 6 6 6] 6 [6 6 6 6 6 6] std-006 [6 6 6 6 6 6] [vec-006 vec-006 vec-006 vec-006 vec-006 vec-006] end-006}
[007][evt]: {beg-007 7 [7 7 7 7 7 7 7 7 7 7] 7 [7 7 7 7 7 7 7] std-007 [7 7 7 7 7 7 7] [vec-007 vec-007 vec-007 vec-007 vec-007 vec-007 vec-007] end-007}
[008][evt]: {beg-008 8 [8 8 8 8 8 8 8 8 8 8] 8 [8 8 8 8 8 8 8 8] std-008 [8 8 8 8 8 8 8 8] [vec-008 vec-008 vec-008 vec-008 vec-008 vec-008 vec-008 vec-008] end-008}
[009][evt]: {beg-009 9 [9 9 9 9 9 9 9 9 9 9] 9 [9 9 9 9 9 9 9 9 9] std-009 [9 9 9 9 9 9 9 9 9] [vec-009 vec-009 vec-009 vec-009 vec-009 vec-009 vec-009 vec-009 vec-009] end-009}
>>> file[./testdata/chain.2.root]
key[000]: tree;1 "my tree title" (TTree)
[000][evt]: {beg-010 10 [10 10 10 10 10 10 10 10 10 10] 0 [] std-010 [] [] end-010}
[001][evt]: {beg-011 11 [11 11 11 11 11 11 11 11 11 11] 1 [11] std-011 [11] [vec-011] end-011}
[002][evt]: {beg-012 12 [12 12 12 12 12 12 12 12 12 12] 2 [12 12] std-012 [12 12] [vec-012 vec-012] end-012}
[003][evt]: {beg-013 13 [13 13 13 13 13 13 13 13 13 13] 3 [13 13 13] std-013 [13 13 13] [vec-013 vec-013 vec-013] end-013}
[004][evt]: {beg-014 14 [14 14 14 14 14 14 14 14 14 14] 4 [14 14 14 14] std-014 [14 14 14 14] [vec-014 vec-014 vec-014 vec-014] end-014}
[005][evt]: {beg-015 15 [15 15 15 15 15 15 15 15 15 15] 5 [15 15 15 15 15] std-015 [15 15 15 15 15] [vec-015 vec-015 vec-015 vec-015 vec-015] end-015}
[006][evt]: {beg-016 16 [16 16 16 16 16 16 16 16 16 16] 6 [16 16 16 16 16 16] std-016 [16 16 16 16 16 16] [vec-016 vec-016 vec-016 vec-016 vec-016 vec-016] end-016}
[007][evt]: {beg-017 17 [17 17 17 17 17 17 17 17 17 17] 7 [17 17 17 17 17 17 17] std-017 [17 17 17 17 17 17 17] [vec-017 vec-017 vec-017 vec-017 vec-017 vec-017 vec-017] end-017}
[008][evt]: {beg-018 18 [18 18 18 18 18 18 18 18 18 18] 8 [18 18 18 18 18 18 18 18] std-018 [18 18 18 18 18 18 18 18] [vec-018 vec-018 vec-018 vec-018 vec-018 vec-018 vec-018 vec-018] end-018}
[009][evt]: {beg-019 19 [19 19 19 19 19 19 19 19 19 19] 9 [19 19 19 19 19 19 19 19 19] std-019 [19 19 19 19 19 19 19 19 19] [vec-019 vec-019 vec-019 vec-019 vec-019 vec-019 vec-019 vec-019 vec-019] end-019}
```

**Comment by sbinet** [comment]
ah, something probably worth skimming a bit:

- [File I/O with C++/ROOT](https://root.cern.ch/root/htmldoc/guides/users-guide/ROOTUsersGuide.html#inputoutput)
- [TTrees with C++/ROOT](https://root.cern.ch/root/htmldoc/guides/users-guide/ROOTUsersGuide.html#trees)

@maloft, this should clear things up a bit about a few ROOT-related concepts. I think.

**Comment by sbinet** [comment]
closed by #213
